shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -619,6 +630,21 @@ - (UITextRange*)rangeEnclosingPosition:(UITextPosition*)position
 }
 
 - (UITextRange*)lineEnclosingPosition:(UITextPosition*)position {
+  // TODO(hellohuanlin): Remove iOS 17 check. The same logic should apply to older versions too.
+  if (@available(iOS 17.0, *)) {
+    // The end of document (with backward affinity) should not be part of any line, since end is
+    // exclusive. This is to fix 2 bugs:
+    // 1. Keyboard freezes when switching languages:
+    // https://github.com/flutter/flutter/issues/134716
+    // 2. Auto correction candidate menu does not show up in iOS 17:
+    // https://github.com/flutter/flutter/issues/132594
+    // The end of document with forward affinity should still return the last line, which is used
+    // by voice control's delete line command.
+    if ([position isEqual:[_textInputView endOfDocument]]) {
```

Comment:
It's non-obvious how this test relates to the comment, since to understand how this is checking affinity the reader has to go look up the definition of `endOfDocument`; please clarify in the comment (or change the logic here to explicitly test the index and direction).

Reply:
> Aren't affinity and text direction actually distinct concepts?

Affinity and text direction are distinct concept. 

> In which case this equality check is relying on a conflation of expression, which we shouldn't do.

Maybe I'm missing something. This checks text position and affinity, and both of them are independent of text direction.

Reply:
Sorry, I forgot that FlutterTextPosition is a subclass of UITextPosition, rather than a distinct thing, so I thought we were comparing our affinity to a UIText system text direction.

So ignore that comment; the earlier comment about the condition being non-obvious still applies though.

Reply:
Since this checks if the position is OOB, shouldn't this be a `>` or `>=` check instead of an equality check?

Reply:
Make sense. I think I will remove the isEqual and put logic here.